solutions/cses-1073.mdx

```diff
@@ -47,7 +60,77 @@ int main() {
 }
 ```
 
-## Implementation 2: Using `multiset`
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class Towers {
+	public static void main(String[] args) throws IOException {
+		Kattio io = new Kattio();
+		int n = io.nextInt();
+		int[] cubes = new int[n];
+		for (int i = 0; i < n; i++) {
+			cubes[i] = io.nextInt();
+		}
+
+		// Stores towers in non-decreasing order
+		List<Integer> towers = new ArrayList<>();
+		for (int i = 0; i < n; i++) {
+			// Upper bound binary search
+			int lo = 0;
+			int hi = towers.size();
+			while (lo < hi) {
+				int mid = (lo + hi) / 2;
+				if (cubes[i] >= towers.get(mid)) lo = mid+1;
+				else hi = mid;
+			}
+			// If there aren't any suitable towers, append new tower to end of
+			// array
+			if (lo == towers.size()) {
+				towers.add(cubes[i]);
+			}
+			// If there exists a satisfying tower, add the cube to that tower
+			// and update the top element of the tower
+			else {
+				towers.set(lo, cubes[i]);
+			}
+		}
+		io.println(towers.size());
+		io.close();
+	}
+
+	// CodeSnip{Kattio}
+}
+````
+</JavaSection>
+
+</LanguageSection>
+
+## Implementation 2: Multiset
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+In this approach, we store the towers using an ordered multiset (which can be
```

Comment:
would it be better to wrap the explanation in `## Explanation` or `## Solution`?

Reply:
So like `Solution 2 - Multiset`?

Reply:
I did implementation because I think the explanation is more geared to how to implement the solution from the beginning

Reply:
nvm solution is probably better